test/e2e/network/traffic_policy.go

```diff
@@ -0,0 +1,453 @@
+// If expectedZone is specified, each request pod will be verified to reside in this zone. Otherwise the zone of the pod is not checked.
+// After a number of responses have been received checkResponses is used to validate which pods handled the request.
+func checkEndpointHost(c clientset.Interface, f *framework.Framework, tr *trafficPolicyTestResources, endpoint string, expectedZone string, nodeName string, checkResponses SeenPodsCheckFunc) {
+	ginkgo.By("creating a client pod for probing the service from " + nodeName)
+	// put a pod in the single node zone to use for curling
+	podName := "curl-from-" + nodeName
+
+	/// turn this into a function which takes endpoint, and expected reply details
+	clientPod := e2epod.NewAgnhostPod(f.Namespace.Name, podName, nil, nil, nil, "serve-hostname")
+	nodeSelection := e2epod.NodeSelection{Name: nodeName}
+	e2epod.SetNodeSelection(&clientPod.Spec, nodeSelection)
+
+	ginkgo.By("Checking access to endpoint " + endpoint)
+
+	cmd := fmt.Sprintf(`date; for i in $(seq 1 3000); do sleep 0.5; echo "Date: $(date) Try: ${i}"; curl -q -s --connect-timeout 2 http://%s/hostname ; echo; done`, endpoint)
+	clientPod.Spec.Containers[0].Command = []string{"/bin/sh", "-c", cmd}
+	clientPod.Spec.Containers[0].Name = clientPod.Name
+	podClient := e2epod.NewPodClient(f)
+	clientPod = podClient.CreateSync(clientPod)
+
+	var logs string
+	seenPods := make(map[string]string)
+	if pollErr := wait.Poll(5*time.Second, e2eservice.KubeProxyLagTimeout, func() (bool, error) {
+		var err error
+		logs, err = e2epod.GetPodLogs(c, f.Namespace.Name, clientPod.Name, clientPod.Name)
+		framework.ExpectNoError(err)
+		framework.Logf("Pod client logs: %s", logs)
+
+		logLines := strings.Split(logs, "\n")
+		if len(logLines) < 6 {
+			framework.Logf("only %d log lines, waiting for at least 6", len(logLines))
+			return false, nil
+		}
+
+		responses := 0
+
+		for i := len(logLines) - 1; i > 0; i-- {
+			if logLines[i] == "" || strings.HasPrefix(logLines[i], "Date:") {
+				continue
+			}
+			destZone, ok := tr.zonesByPod[logLines[i]]
+			if !ok {
+				framework.Logf("could not determine dest zone from log line: %s", logLines[i])
+				return false, nil
+			}
+			seenPods[logLines[i]] = destZone
+			// if a expectedZone is provided, verify traffic stays within
+			if expectedZone != "" && expectedZone != destZone {
+				framework.Logf("expected request from %s to stay in %s zone, delivered to %s zone", clientPod.Name, expectedZone, destZone)
+				return false, nil
+			}
+			responses++
+			if responses >= 10 {
+				if checkResponses(tr, seenPods) {
+					return true, nil
+				} else {
+					return false, fmt.Errorf("invalid responses")
+				}
+			}
+		}
+		return false, nil
+	}); pollErr != nil {
+		framework.Failf("expected 10 consecutive requests from %s to stay in zone %s within %v, stdout: %v", clientPod.Name, expectedZone, e2eservice.KubeProxyLagTimeout, logs)
+	}
+	podClient.DeleteSync(podName, metav1.DeleteOptions{}, 10*time.Second)
```

Comment:
this will be cleaned by the test when it finish

Reply:
Yes, but in this `ginko.It` block the client container may be placed differently when internal vs external request endpoint is checked with `checkEndpointHost`

For example line 364 and 368.

Basically each configuration that's tested creates the nodeport service. Then to check which pod handles requests, the test creates a client pod and tries a request to an 'external' endpoint, then creates a pod to check the internal endpoint. 

This delete stands out, for sure. If it's preferable I could rework this to always create two client pods. It would make the test faster. Let me know what you think.

Reply:
sorry, I didn't check the PR carefully, will try to review better later, creating and execing pods is expensive and increase the risk of failures on the test, if we can reuse it much better ...

Reply:
nevermind, I got your point now, it is fine, scratch this comment

Reply:
In the latest update I changed the tests to no longer delete these pods. Instead I create them after the service and did what you suggest and used trafficPolicyTestResources struct to keep the internal and external client pod names. Seems cleaner, thanks! Resolving this comment.